### DISCUSSION TITLE
How to configure multiple data sources on startup?

### DISCUSSION THREAD

**Discussion by paymog** [hasura/graphql-engine#9665] [Question] [answered]
I'm currently running the community edition of hasura in my infrastructure and I'm using `HASURA_GRAPHQL_DATABASE_URL`. We're now upgrading our infra and are examining ways of configuring multiple data sources. 

I found [these docs](https://hasura.io/docs/latest/databases/database-config/index/#connect-a-database-via-the-hasura-cli-or-api) but I can't quite figure out how to wire up a config file like that. I notice that the title of the section mentions using the cli and api. I've also searched for how to "point" to a config file at startup with the docker image but haven't been successful.

Is there a way to configure hasura at startup with more than 1 data source?

**Comment by SamirTalwar** [comment, marked as answer]
Data sources are configured at runtime by adding them using the metadata API. There's no way of pointing Hasura GraphQL Engine to a configuration file at startup because the configuration is stored in the metadata database. However, you can use the CLI to apply a configuration to _a running HGE instance_.

As you've been using `HASURA_GRAPHQL_DATABASE_URL`, Hasura is configured _the first time it starts up_ with that database as both the metadata DB and the source DB named "default". When you start it again later, the URL will be used for the metadata database _only_. HGE will then read the metadata, which already contains the same URL for the "default" source DB.

In order to add a new data source to an existing HGE instance, you can use [the `pg_add_source` API endpoint](https://hasura.io/docs/latest/api-reference/metadata-api/source/) (or equivalent for another database), or use the Hasura Console to the same thing (under the "Data" tab).

Once you've done this, you can set up the Hasura CLI and run `hasura metadata export` to download the current metadata as a set of YAML files. You can then run `hasura metadata apply` to re-apply this configuration to a running HGE instance.

**Reply by cmartin616** [reply]
Additionally, you can specify any plaintext named environment variable (I'd recommend not starting HASURA_*) for each of your databases.  For example:

- PG_APP_DB_X: [conn_string]
- PG_APP_DB_X_READ_REPLICA: [conn_string]
- PG_APP_DB_Y: [conn_string]
- ...

Then configure your data sources to use the environment variable rather than a hard-coded value in the metadata.  This is also essential for devops/tier promotions because you need to be able to change the value of the database connection string for each tier _without_ altering the metadata.  This allows for the same metadata to be used in each environment where each environment has different environment variables.
